File: backend/app/models/tailoring.py
from __future__ import annotations

from typing import Literal

from pydantic import Field

from app.models.common import ApiModel

TailoredOperation = Literal["rewrite", "reorder", "emphasize", "remove"]
TailoredClassification = Literal["REPHRASED", "REORDERED", "EMPHASIZED", "REMOVED", "NEW_CLAIM"]
ReviewStatusValue = Literal["proposed", "approved", "rejected"]


class TailoredChangeOutput(ApiModel):
    section: str = Field(min_length=1, max_length=120)
    operation: TailoredOperation
    before: str = Field(default="", max_length=2000)
    after: str = Field(min_length=1, max_length=2000)
    classification: TailoredClassification
    reason: str = Field(min_length=1, max_length=1000)
    source_fact_ids: list[str] = Field(default_factory=list)


class TailoringAgentOutput(ApiModel):
    resume_changes: list[TailoredChangeOutput] = Field(min_length=1, max_length=60)
    cover_letter_paragraphs: list[str] = Field(min_length=1, max_length=10)
    cover_letter_source_fact_ids: list[str] = Field(default_factory=list)


class DocumentTailorRequest(ApiModel):
    profile_id: str
    job_id: str


class DocumentChangeOutput(ApiModel):
    id: str
    section: str
    operation: TailoredOperation
    before: str
    after: str
    classification: TailoredClassification
    reason: str
    source_fact_ids: list[str]
    status: ReviewStatusValue


class ResumeDocumentOutput(ApiModel):
    id: str
    status: str
    changes: list[DocumentChangeOutput]


class CoverLetterDocumentOutput(ApiModel):
    id: str
    status: str
    paragraphs: list[str]
    source_fact_ids: list[str]


class DocumentTailorResponse(ApiModel):
    operation_id: str
    profile_id: str
    job_id: str
    resume: ResumeDocumentOutput
    cover_letter: CoverLetterDocumentOutput
    provider: str
    model: str
    prompt_version: str


class DocumentChangeReviewRequest(ApiModel):
    status: Literal["approved", "rejected"]


class DocumentChangeReviewResponse(ApiModel):
    id: str
    status: ReviewStatusValue

File: webapp/src/schemas/tailoring.ts
import { z } from 'zod';

export const documentChangeSchema = z.object({
  id: z.string(),
  section: z.string(),
  operation: z.enum(['rewrite', 'reorder', 'emphasize', 'remove']),
  before: z.string(),
  after: z.string(),
  classification: z.enum(['REPHRASED', 'REORDERED', 'EMPHASIZED', 'REMOVED', 'NEW_CLAIM']),
  reason: z.string(),
  source_fact_ids: z.array(z.string()),
  status: z.enum(['proposed', 'approved', 'rejected']),
  supported: z.boolean(),
});

export const resumeItemSchema = z.object({
  id: z.string(),
  label: z.string(),
  value: z.string(),
  source_fact_ids: z.array(z.string()),
});

export const resumeSectionSchema = z.object({
  id: z.string(),
  title: z.string(),
  items: z.array(resumeItemSchema),
});

export const canonicalResumeSchema = z.object({
  name: z.string(),
  email: z.string().nullable(),
  contact: z.record(z.string(), z.string()),
  sections: z.array(resumeSectionSchema),
  template_id: z.string(),
  page_size: z.string(),
});

export const resumeDocumentSchema = z.object({
  id: z.string(),
  status: z.string(),
  changes: z.array(documentChangeSchema),
  source: canonicalResumeSchema,
  current: canonicalResumeSchema,
});

export const coverLetterDocumentSchema = z.object({
  id: z.string(),
  status: z.string(),
  paragraphs: z.array(z.string()),
  source_fact_ids: z.array(z.string()),
});

export const documentTailorResponseSchema = z.object({
  operation_id: z.string(),
  profile_id: z.string(),
  job_id: z.string(),
  resume: resumeDocumentSchema,
  cover_letter: coverLetterDocumentSchema,
  provider: z.string(),
  model: z.string(),
  prompt_version: z.string(),
});

export const documentChangeReviewResponseSchema = z.object({
  id: z.string(),
  status: z.enum(['proposed', 'approved', 'rejected']),
});

export const generatedResumeSchema = z.object({
  id: z.string(),
  job_id: z.string(),
  title: z.string(),
  company: z.string().nullable(),
  source_url: z.string().nullable(),
  source: canonicalResumeSchema,
  current: canonicalResumeSchema,
  changes: z.array(documentChangeSchema),
});

export type DocumentChange = z.infer<typeof documentChangeSchema>;
export type DocumentTailorResponse = z.infer<typeof documentTailorResponseSchema>;
export type CanonicalResume = z.infer<typeof canonicalResumeSchema>;
export type GeneratedResume = z.infer<typeof generatedResumeSchema>;
